drivers/storage/portworx/portworx.go

```diff
@@ -168,6 +182,239 @@ func (p *portworx) DeleteStorage(storageCluster *corev1alpha1.StorageCluster) (*
 	}, nil
 }
 
+func (p *portworx) UpdateStorageClusterStatus(
+	cluster *corev1alpha1.StorageCluster,
+) error {
+	clientConn, err := p.getPortworxClient(cluster)
+	if err != nil {
+		return err
+	}
+
+	clusterClient := api.NewOpenStorageClusterClient(clientConn)
+	pxCluster, err := clusterClient.InspectCurrent(context.TODO(), &api.SdkClusterInspectCurrentRequest{})
+	if err != nil {
+		p.sdkConn = nil
+		return fmt.Errorf("failed to inspect cluster: %v", err)
+	} else if pxCluster.Cluster == nil {
+		return fmt.Errorf("empty ClusterInspect response")
+	}
+
+	cluster.Status.Phase = string(mapClusterStatus(pxCluster.Cluster.Status))
+	cluster.Status.ClusterName = pxCluster.Cluster.Name
+	cluster.Status.ClusterUID = pxCluster.Cluster.Id
+
+	return p.updateNodeStatuses(clientConn, cluster)
+}
+
+func (p *portworx) updateNodeStatuses(
+	clientConn *grpc.ClientConn,
+	cluster *corev1alpha1.StorageCluster,
+) error {
+	nodeClient := api.NewOpenStorageNodeClient(clientConn)
+	nodeEnumerateResponse, err := nodeClient.Enumerate(context.TODO(), &api.SdkNodeEnumerateRequest{})
+	if err != nil {
+		return fmt.Errorf("failed to enumerate nodes: %v", err)
+	}
+
+	ownerRef := metav1.NewControllerRef(cluster, controllerKind)
+	for _, nodeID := range nodeEnumerateResponse.NodeIds {
+		nodeResp, err := nodeClient.Inspect(context.TODO(), &api.SdkNodeInspectRequest{NodeId: nodeID})
+		if err != nil {
+			logrus.Warnf("Failed to inspect node %v: %v", nodeID, err)
+			continue
+		} else if nodeResp.Node == nil {
+			logrus.Warnf("Empty NodeInspect response for nodeID %v: %v", nodeID, err)
+			continue
+		}
+
+		if nodeResp.Node.SchedulerNodeName == "" {
+			k8sNode, err := k8s.Instance().SearchNodeByAddresses(
+				[]string{nodeResp.Node.DataIp, nodeResp.Node.MgmtIp, nodeResp.Node.Hostname},
+			)
+			if err != nil {
+				logrus.Warnf("Unable to find the k8s node name for nodeID %v: %v", nodeID, err)
+				continue
+			}
+			nodeResp.Node.SchedulerNodeName = k8sNode.Name
+		}
+
+		nodeStatus := &corev1alpha1.StorageNodeStatus{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:            nodeResp.Node.SchedulerNodeName,
+				Namespace:       cluster.Namespace,
+				OwnerReferences: []metav1.OwnerReference{*ownerRef},
+				Labels:          p.GetSelectorLabels(),
+			},
+			Status: corev1alpha1.NodeStatus{
+				NodeUID: nodeResp.Node.Id,
+				Network: corev1alpha1.NetworkStatus{
+					DataIP: nodeResp.Node.DataIp,
+					MgmtIP: nodeResp.Node.MgmtIp,
+				},
+				// TODO: Add a human readable reason with the status
+				Conditions: []corev1alpha1.NodeCondition{
+					{
+						Type:   corev1alpha1.NodeState,
+						Status: mapNodeStatus(nodeResp.Node.Status),
+					},
```

Comment:
We need StorageState as well here

Reply:
We don't have a way to get storage status right now. Once we have the internal SDK we can add that.